### ISSUE TITLE
new.stats.jenkins.io slow to load

### ISSUE THREAD

**Issue by shlomomdahan** [jenkins-infra/helpdesk#4177]
### Service(s)

Helpdesk

### Summary

ref:
 - #4132 

This site is being built as part of GSoC. While the site performs well locally, the deployment links and the temporary [new.stats.jenkins.io](new.stats.jenkins.io) take a very long time to load (15-60 seconds or more).

@lemeurherve @gounthar @krisstern @Vandit1604 

### Reproduction steps

N/A

**Comment by dduportal** [comment]
Hi @shlomomdahan , at first (really quick and superficial analysis), I see that the website loads a 73Mb javascript asset at https://new.stats.jenkins.io/assets/index-Bz5xGm9h.js.

I'm not sure what it does (it's full of JS code) but that is not acceptable to have a file of this size. Do you have the same file when running (fast) locally? Even if you disable cache?

If you see other slowness, you need to diagnose it on your side as we are missing information here:

- What is the web resource slow to load on your case? Can you reproduce the slowness using `curl`?
- Do you have web inspector view?
- Can you detail what is "slow" in your case precisely?

**Comment by shlomomdahan** [comment]
Hi @dduportal,

Thank you for highlighting the issue. By using the Webpack bundle analyzer, I managed to reduce the JS asset size by half. I identified and removed a redundant large import, which significantly improved the load time.

Currently, the app employs a glob import at build time to load all the data, handled by the Vite framework. However, it appears that to achieve further speed improvements, we'll need to change the file import method.

I appreciate your time and feedback on this matter!

Thanks!

**Comment by dduportal** [comment]
@shlomomdahan thanks for the first step on these efforts. The JS bundle is now around 61Mb (it was 73 Mb when this issue was raised).

Still some ways to go: it sounds crazy using that much of JS: I guess that's because of the frameworks. If you have a way to minify and shrink even more as the next logical steps (first: make something which works, then optimizes) that would be really useful.

On the infrastructure side, we've enabled both [`gzip`](https://github.com/jenkins-infra/kubernetes-management/pull/5442) and [`brotli`](https://github.com/jenkins-infra/kubernetes-management/pull/5444) compression which gives the following results with the current ~61 Mb JS bundle:

- With an HTTP client using `gzip`: ~13Mb
- With an HTTP client using `brotli` ("br"): ~10 Mb

=> this should help also help lot, without removing the need to optimize the generated assets. Please note that this would be a requirement for opening to production (e.g. shrink this asset to have a final compressed file under 5 Mb) BUT this might not be the immediate step for you to run (please check with your GSoC mentor(s).

**Comment by dduportal** [comment]
@shlomomdahan Can you confirm the current status is good enough for you by closing this issue, or comment with more details if you expect more actions from the infra team.

**Comment by shlomomdahan** [comment]
@dduportal

Thank you for helping with this and providing all of the above info. I see the website loading much faster now (around 3-4 seconds), which is a great improvement.

I will see if my mentors have any suggestions for optimizing the asset size to <5MB to ensure it's acceptable for production. Thanks again for your support!

**Comment by shlomomdahan** [comment]
![Screenshot 2024-07-22 at 12 12 39 PM](https://github.com/user-attachments/assets/59d2a3ec-df2a-47e8-b257-d5df2e762619)

Hi @dduportal,

I think I may have misunderstood the above. The huge bundle size is not actually due to large amounts of JS code, but rather the large amount of data we are 'eagerly' loading up into the application at build time.

@lemeurherve suggested at an earlier stage of the project to load and incorporate all the data into the app during build time, so that we could serve the user a static site once deployed. This made a lot of sense because making, for example, 2000+ Axios requests to load all the plugin data was not feasible and caused its own performance issues.

So, at the current state, the below code loads up all the data during build and populates the entire site with the data. But since this is the case, I don't really see how it's possible to reduce the bundle size simply because we are using so much data.

```
const csvFiles = import.meta.glob('../data/infra-statistics/jenkins-stats/svg/*.csv', {
    eager: true,
    query: '?raw',
    import: 'default',
}) as Record<string, string>
const jsonFiles = import.meta.glob('../data/infra-statistics/plugin-installation-trend/*.json', {
    eager: true,
    query: '?raw',
    import: 'default',
}) as Record<string, string>

const csvFileMapping: Record<string, string> = {}
const jsonFileMapping: Record<string, string> = {}

for (const path in csvFiles) {
    const fileName = path.split('/').pop()?.replace('.csv', '')
    if (fileName) {
        csvFileMapping[fileName] = csvFiles[path]
    }
}

for (const path in jsonFiles) {
    const fileName = path.split('/').pop()?.replace('.stats.json', '')
    if (fileName) {
        jsonFileMapping[fileName] = jsonFiles[path]
    }
}

export { csvFileMapping, jsonFileMapping }

```

Given this context, I would appreciate any insights on how we could manage this data loading strategy more efficiently or alternative approaches to handle large datasets in our build process.

Thanks!

**Comment by dduportal** [comment]
> ![Screenshot 2024-07-22 at 12 12 39 PM](https://private-user-images.githubusercontent.com/64103471/351037090-59d2a3ec-df2a-47e8-b257-d5df2e762619.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2Njc0NjQsIm5iZiI6MTcyMTY2NzE2NCwicGF0aCI6Ii82NDEwMzQ3MS8zNTEwMzcwOTAtNTlkMmEzZWMtZGYyYS00N2U4LWIyNTctZDVkZjJlNzYyNjE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDE2NTI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0NzU1NzY5ZGYwZTkyY2Q4ZTRiOGY5N2M2ZjFmNTFhMmNiNTVkMDQ4ODIzN2RlOTNlMGQ2Njg4ODU3ZTk3MTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Ou70UVaVnHK1iLfy3JQDMSdGT7QgAbH6MwSCRjleXJI)
> 
> Hi @dduportal,
> 
> I think I may have misunderstood the above. The huge bundle size is not actually due to large amounts of JS code, but rather the large amount of data we are 'eagerly' loading up into the application at build time. (above image shows the rollup visualizer)
> 
> @lemeurherve suggested at an earlier stage of the project to load and incorporate all the data into the app during build time, so that we could serve the user a static site once deployed. This made a lot of sense because making, for example, 2000+ Axios requests to load all the plugin data was not feasible and caused its own performance issues.
> 
> So, at the current state, the below code loads up all the data during build and populates the entire site with the data. But since this is the case, I don't really see how it's possible to reduce the bundle size simply because we are using so much data.
> 
> ```
> const csvFiles = import.meta.glob('../data/infra-statistics/jenkins-stats/svg/*.csv', {
>     eager: true,
>     query: '?raw',
>     import: 'default',
> }) as Record<string, string>
> const jsonFiles = import.meta.glob('../data/infra-statistics/plugin-installation-trend/*.json', {
>     eager: true,
>     query: '?raw',
>     import: 'default',
> }) as Record<string, string>
> 
> const csvFileMapping: Record<string, string> = {}
> const jsonFileMapping: Record<string, string> = {}
> 
> for (const path in csvFiles) {
>     const fileName = path.split('/').pop()?.replace('.csv', '')
>     if (fileName) {
>         csvFileMapping[fileName] = csvFiles[path]
>     }
> }
> 
> for (const path in jsonFiles) {
>     const fileName = path.split('/').pop()?.replace('.stats.json', '')
>     if (fileName) {
>         jsonFileMapping[fileName] = jsonFiles[path]
>     }
> }
> 
> export { csvFileMapping, jsonFileMapping }
> ```
> 
> Given this context, I would appreciate any insights on how we could manage this data loading strategy more efficiently or alternative approaches to handle large datasets in our build process.
> 
> Thanks!

Hi @shlomomdahan , as I never used VueJS or Webpack for that matter so I'm not sure.

However, if I open the big JS file at https://new.stats.jenkins.io/assets/index-uEElUgy9.js with my text editor, I see the following elements:

- There are around 4Mb of only carriage returns (is there an "inline" option to have a 1 line file)?
- I do not understand the rationale of putting CSS, JS and data in the same file: is there a way to split these in 3 different assets?
- Regarding the data, it feels like the current formats are not ideals: both CSV or JSON looks really heavy. Isn't there a binary format which could be used?

**Comment by timja** [comment]
You can bundle the data up so that it's not 2000 requests but instead just one or one per report so that you only need the data that you need.

i.e. I absolutely should not need to load a 10.4 MB script to load the home page which has no data on it: https://new.stats.jenkins.io/

It can still be a static site without loading the data directly in the main bundle.

I would suggest starting by splitting the bundle into what you need ideally per report and only loading the data you need.
Minifying further as suggest by @dduportal would be good. The scripts that I looked at appear to be minified but the data isn't.

>  Isn't there a binary format which could be used?

I wouldn't jump there first it will generally be a lot harder to debug and is likely unnecessary for this.

**Comment by shlomomdahan** [comment]
Hi @timja @dduportal,

Based on your recommendations, I successfully changed the data fetching method to be dynamic vs fetching it all during the build phase. As a result, the bundle size now appears to be <5mb on my end (probably even smaller with the compression implemented in [#5442](https://github.com/jenkins-infra/kubernetes-management/pull/5442). The load time has improved drastically as well.

Thank you for your help and insights on this!

**Comment by gounthar** [comment]
What an achievement, congrats! 🥳

**Comment by dduportal** [comment]
> Hi @timja @dduportal,
> 
> Based on your recommendations, I successfully changed the data fetching method to be dynamic vs fetching it all during the build phase. As a result, the bundle size now appears to be <5mb on my end (probably even smaller with the compression implemented in [#5442](https://github.com/jenkins-infra/kubernetes-management/pull/5442). The load time has improved drastically as well.
> 
> Thank you for your help and insights on this!

The result is impressive, congratulations!
